### PR TITLE
[move] Added inverse traversal CFG

### PR DESCRIPTION
**Pull request by tnowacki**
## Motivation

- Added a CFG that traverses blocks and commands in inverse order
  - There is a fake terminal node added as the root/start for the reverse traversal 
  - The predecessors of the fake node are any node without a normal successor 
  - Additionally, any infinite loop without a break/return/abort will be a predecessor of this fake terminal node
- Will be used later for liveness analysis

## Test Plan

- no new tests yet
- assertions are tested on construction
- New tests will be added after usage

### CODE REVIEW

File: language/move-lang/src/cfgir/cfg.rs [hunk 1]
```diff
@@ -70,31 +87,145 @@ impl<'a> BlockCFG<'a> {
...
+    pub fn block(&self, label: Label) -> &BasicBlock {
+        self.blocks.get(&label).unwrap()
+    }
+
+    pub fn block_mut(&mut self, label: Label) -> &mut BasicBlock {
+        self.blocks.get_mut(&label).unwrap()
     }
+}
+
+impl<'a> Drop for ReverseBlockCFG<'a> {
```
Comment (tnowacki) [inline language/move-lang/src/cfgir/cfg.rs]: Not sure if is is the best way to do this, but it seemed to be the least brittle.

Comment (sblackshear) [reply language/move-lang/src/cfgir/cfg.rs]: I think taking `forward_cfg: &'a BlockCFG` above and cloning `forward_cfg.blocks` would be even less brittle. That is, unless you actually want to mutate the `blocks` of the forward CFG later, which I could see being useful for optimization.

This `Drop` trick is interesting as a way of building a temporary "view" in Rust. You could imagine mutating `successor_map` and `predecessor_map` above instead of cloning them, then undoing those mutations in `Drop` as well.

File: language/move-lang/src/hlir/translate.rs [hunk 2]
```diff
@@ -521,6 +528,17 @@ fn statement(context: &mut Context, result: &mut Block, e: T::Exp) {
...
 
+fn statement_loop_body(context: &mut Context, body: T::Exp) -> (Block, bool) {
+    let old_has_return_abort = context.has_return_abort;
+    context.has_return_abort = false;
+    let mut loop_block = Block::new();
+    let el = maybe_exp(context, &mut loop_block, None, body);
+    ignore_and_pop(context, &mut loop_block, true, el);
+    let has_return_abort = context.has_return_abort;
+    context.has_return_abort = context.has_return_abort || old_has_return_abort;
+    (loop_block, has_return_abort)
```
Comment (sblackshear) [inline language/move-lang/src/hlir/translate.rs]: Can the caller compute the value of `has_return_abort` instead? The dual purpose of creating block and computing this bool makes this function a bit hard to follow.

File: language/move-lang/src/cfgir/cfg.rs [hunk 3]
```diff
@@ -70,31 +87,145 @@ impl<'a> BlockCFG<'a> {
...
+            for (lbl, block) in blocks.iter() {
+                let last_cmd = block.back().unwrap();
+                if last_cmd.value.is_divergent() {
+                    end_blocks.insert(*lbl);
+                }
+            }
+            end_blocks
+        };
+
+        let mut successor_map = forward_predecessor.clone();
```
Comment (sblackshear) [inline language/move-lang/src/cfgir/cfg.rs]: Nit: might be less confusing to keep the usual interpretation of predecessors and successors for as long as possible (e.g., `predecessor_map.instert(terminal, end_blocks)` below instead), then do `successor_map: predecessor_map, predecessor_map: succesor_map` when you actually build the struct.

File: language/move-lang/src/cfgir/absint.rs [hunk 4]
```diff
@@ -153,12 +153,12 @@ pub trait AbstractInterpreter: TransferFunctions {
 
     fn execute_block(
         &mut self,
-        cfg: &BlockCFG,
+        cfg: &mut dyn CFG,
```
Comment (sblackshear) [inline language/move-lang/src/cfgir/absint.rs]: Why does this need to be mutable?

File: language/move-lang/src/cfgir/absint.rs [hunk 5]
```diff
@@ -77,12 +77,12 @@ pub trait TransferFunctions {
...
     /// information allows clients to detect the end of a basic block and special-case appropriately
     /// (e.g., normalizing the abstract state before a join).
-    fn execute(&mut self, pre: &mut Self::State, command: &Command) -> Errors;
+    fn execute(&mut self, pre: &mut Self::State, command: &mut Command) -> Errors;
 }
 
 pub trait AbstractInterpreter: TransferFunctions {
     /// Analyze procedure local@function_view starting from pre-state local@initial_state.
-    fn analyze_function(&mut self, cfg: &BlockCFG, initial_state: Self::State) -> Errors {
+    fn analyze_function(&mut self, cfg: &mut dyn CFG, initial_state: Self::State) -> Errors {
```
Comment (sblackshear) [inline language/move-lang/src/cfgir/absint.rs]: Similar question here and in `execute` about mutability.

File: language/move-lang/src/cfgir/absint.rs (hunk 4 above)
Comment (tnowacki) [reply language/move-lang/src/cfgir/absint.rs]: I had originally assumed I would be using the abstract interpreter frame work to insert the releases. But I think that is not true

File: language/move-lang/src/hlir/translate.rs (hunk 2 above)
Comment (tnowacki) [reply language/move-lang/src/hlir/translate.rs]: > Can the caller compute the value of has_return_abort instead?

No, I need to update the state correctly, and return the proper value. I'd rather not make the caller also update the context.

File: language/move-lang/src/cfgir/ast.rs [hunk 6]
```diff
@@ -229,6 +229,14 @@ impl Command_ {
         }
     }
 
+    pub fn is_divergent(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply language/move-lang/src/cfgir/ast.rs]: `is_exit` is good. I struggled naming this and picked something hoping someone would comment :)

File: language/move-lang/src/cfgir/cfg.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-lang/src/cfgir/cfg.rs]: > That is, unless you actually want to mutate the blocks of the forward CFG later, which I could see being useful for optimization.

Originally, I had planned on mutating, but I'm unsure what I'll do exactly there now.

> This Drop trick is interesting as a way of building a temporary "view" in Rust. You could imagine mutating successor_map and predecessor_map above instead of cloning them, then undoing those mutations in Drop as well.

I am afraid of over relying on drop too much, but maybe it will be fine since I will be taking a mutable reference (and thus taking ownership). It's something to ponder, but I might stick with using `Drop` as little as possible right now

File: language/move-lang/src/cfgir/absint.rs (hunk 4 above)
Comment (sblackshear) [reply language/move-lang/src/cfgir/absint.rs]: Yeah, mutating code while trying to compute a fixed point over it sounds ill-advised...  A postprocessing step would probably be cleaner.

File: language/move-lang/src/cfgir/cfg.rs (hunk 3 above)
Comment (tnowacki) [reply language/move-lang/src/cfgir/cfg.rs]: Done

File: language/move-lang/src/cfgir/cfg.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-lang/src/cfgir/cfg.rs]: I actually ended up taking references to both, and then cleaning up with `Drop`. The implementation ended up looking pretty nice
